src/Conjure/UI/Model.hs

```diff
@@ -1848,6 +1836,128 @@ rule_DotLtLeq = "generic-DotLtLeq" `namedRule` theRule where
+                    _ -> bug $ "epilogue: flattenLex: isn't defined for this fellow..."
+                        <+> vcat [pretty a, pretty ta, stringToDoc $ show a]
+                _ -> bug $ "epilogue: flattenLex: isn't defined for this fellow..."
+                    <+> vcat [pretty a, pretty ta, stringToDoc $ show a]
+            Op _ -> do
+              (Single oName, o) <- quantifiedVar
+              flatten $ Comprehension o [ComprehensionLetting oName a]
+            _ -> do
+              ps <- sequence $ (\(i,_) -> do
+                                  (Single nm, tm) <- quantifiedVar
+                                  return (i,nm,tm)) <$> (zip [1..] ts)
+              let lts = (\(i,nm,tm) -> ComprehensionLetting nm [essence| &a[&i] |]) <$> ps
+                  tup = AbstractLiteral $ AbsLitTuple $ (\(_,_,tm) -> tm) <$> ps
+              flatten $ Comprehension tup lts
+        _ ->
+          case a of
+            AbstractLiteral x -> do
+              case x of
+                AbsLitMatrix _ xs -> do
+                  fxs <- sequence (flatten <$> xs)
+                  let flatxs = fromList fxs
+                  return [essence| flatten(&flatxs) |]
+                _ -> bug $ "epilogue: flattenLex: isn't defined for this abslit fellow..."
+                    <+> vcat [pretty a, pretty ta, stringToDoc $ show a]
+            Constant c ->
+              case c of
+                ConstantAbstract ca ->
+                  case ca of
+                    AbsLitMatrix _ [] ->
+                      return [essence| ([] : `matrix indexed by [int()] of int`) |]
+                    AbsLitMatrix _ xs -> do
+                      fxs <- sequence (flatten <$> (Constant <$> xs))
+                      let flatxs = fromList fxs
+                      return [essence| flatten(&flatxs) |]
+                    _ -> bug $ "epilogue: flattenLex: isn't defined for this fellow..."
+                        <+> vcat [pretty a, pretty ta, stringToDoc $ show a]
+                TypedConstant tc _ -> flatten (Constant tc)
+                _ -> bug $ "epilogue: flattenLex: isn't defined for this constant fellow."
+                    <+> vcat [pretty a, pretty ta, stringToDoc $ show a]
+            Op _ -> do
+              (Single oName, o) <- quantifiedVar
+              flatten $ Comprehension o [ComprehensionLetting oName a]
+            Reference nm ex ->
+                  bug $ "epilogue: flattenLex: flatten not defined for this referenced fellow."
+                     <+> vcat [stringToDoc (show a)
+                              ,"reference:" <+> stringToDoc (show nm)
+                              ,"fellow:" <+> stringToDoc (show ex)]
+            Comprehension body gocs -> do
+              fbody <- flatten body
+              let comp = Comprehension fbody gocs
+              return [essence| flatten(&comp) |]
+            _ -> bug $ "epilogue: flattenLex: isn't defined for this expression fellow..."
```

Comment:
this probably shouldn't say epilogue, right?

I see all of the "ugliness" is contained within this rule, which is nicer than what we had before.

Reply:
Yes, the epilogue is cruft, will remove.

I think the ugliness could be polished a little. It is ugly since there is the implicit assumption that symmetryOrdering reduces to only Matrix/Tuple/Comprehension/Int types so everything looks scarily partial :)